### PR TITLE
feat(internal/gensupport): wrap retry failures with context and prev error

### PR DESCRIPTION
**Pull request by matheuscscp**
I have a use case in which my context is cancelling a `sendAndRetry()` operation, and the error being returned is the error returned by the Google API in the previous attempt of `sendAndRetry()` instead of `ctx.Err()`, which is misleading because I was investigating why the Google Storage Go library was giving up on a 503 error (which is not expected according to the retry/idempotency docs: https://cloud.google.com/storage/docs/retry-strategy).

closes: #1685

### CODE REVIEW

**Comment by codyoss** [comment]
@matheuscscp could you please file an issue for discussion before I take a look at the PR. Thanks

**Comment by matheuscscp** [comment]
> @matheuscscp could you please file an issue for discussion before I take a look at the PR. Thanks

@codyoss here it is: https://github.com/googleapis/google-api-go-client/issues/1685

sorry for not starting with that!

File: internal/gensupport/send.go
Comment (quartzmo) [inline internal/gensupport/send.go:100]: Would another possible solution to #1685 be simply to remove this entire block, and let the context raise the deadline exceeded error on its own?

Comment (matheuscscp) [reply internal/gensupport/send.go:100]: do you mean just `return nil, ctx.Err()`? otherwise, where would you suggest to put something equivalent to `case <-time.After(pause):` (the backoff)?

Comment (quartzmo) [reply internal/gensupport/send.go:100]: @tritone Can you comment on this?

Comment (tritone) [reply internal/gensupport/send.go:100]: Hey @quartzmo , thanks for flagging. I think the right answer is to return a wrapped error that includes both the context cancelation and the final error from the service, if available -- otherwise it can be pretty hard to debug either way. We implemented this in google-cloud-go: https://github.com/googleapis/google-cloud-go/blob/0da9ab0831540569dc04c0a23437b084b1564e15/internal/retry.go#L36

I can make a PR to implement this if folks agree.

Comment (tritone) [reply internal/gensupport/send.go:100]: (Also I see that this PR already basically takes that approach -- happy to work with you on completing it yourself if you want; the main change I would ask for is to implement a custom error type that implements `Is(error) bool` so that we can preserve the validity of calls to `errors.Is` in user code.

Comment (matheuscscp) [reply internal/gensupport/send.go:100]: yep I'd like to fix that as you suggested, I will do it tomorrow

Comment (tritone) [reply internal/gensupport/send.go:100]: sounds good, thank you!

Comment (matheuscscp) [reply internal/gensupport/send.go:121]: it's there, PTAL. I basically copied the `wrappedCallErr` struct from your other repo with the exception that I did `Is()` like this:

```
func (e wrappedCallErr) Is(target error) bool {
	return errors.Is(e.ctxErr, target) || errors.Is(e.wrappedErr, target)
}
```

basically leveraging `errors.Is()` instead of using plain `==` comparison. if that's too much pls let me know!

I also seized the opportunity to use this wrapper in all the other occurrences of some ctx+svc error

File: internal/gensupport/send.go
Comment (tritone) [inline internal/gensupport/send.go:120]: This comment is no longer accurate, I would rephrase to something like:

```
// If we got an error and the context has been canceled, return an error acknowledging
// both the context cancelation and the service error.
```

File: internal/gensupport/send.go
Comment (tritone) [inline internal/gensupport/send.go:67]: Let's not make this change in `send()` for now. Reasons:

1. The case that the call returned a non-context error but the context is expired/canceled is not a very common case without retries being in the picture. And the string from `wrappedCallErr` is specific to retries being at the root of the issue.
2. `send` is much more widely used than `sendAndRetry` including by the manual clients built on top of this library. I would want to audit these to make sure they can handle this error (meaning that they are upgraded to use `errors.Is` rather than checking for equality) if they have error handling logic at that layer.

**Review by tritone — COMMENTED** [review]
Generally looks good, aside from change to `send` which I think we should hold off on.

File: internal/gensupport/send.go
Comment (matheuscscp) [reply internal/gensupport/send.go]: fixed!

File: internal/gensupport/send.go
Comment (matheuscscp) [reply internal/gensupport/send.go]: fixed!

**Review by tritone — APPROVED** [review]
LGTM. @codyoss could you also give this a look before we merge?

**Pull request merged**
